neo/SmartContract/StateReader.cs

```diff
@@ -948,8 +948,23 @@ protected virtual bool Storage_Find(ExecutionEngine engine)
                 StorageContext context = _interface.GetInterface<StorageContext>();
                 if (!CheckStorageContext(context)) return false;
                 byte[] prefix = engine.EvaluationStack.Pop().GetByteArray();
-                prefix = context.ScriptHash.ToArray().Concat(prefix).ToArray();
-                StorageIterator iterator = new StorageIterator(Storages.Find(prefix).GetEnumerator());
+                byte[] prefix_key;
+                using (MemoryStream ms = new MemoryStream())
+                {
+                    int index = 0;
+                    int remain = prefix.Length;
+                    while (remain >= 16)
+                    {
+                        ms.Write(prefix, index, 16);
+                        ms.WriteByte(0);
+                        index += 16;
+                        remain -= 16;
+                    }
+                    if (remain > 0)
+                        ms.Write(prefix, index, remain);
+                    prefix_key = context.ScriptHash.ToArray().Concat(ms.ToArray()).ToArray();
+                }
+                StorageIterator iterator = new StorageIterator(Storages.Find(prefix_key).Where(p => p.Key.Key.Take(prefix.Length).SequenceEqual(prefix)).GetEnumerator());
```

Comment:
Is the `Where` part necessary? Since the prefix does not write any additional padding now

Reply:
Yes, it's necessary. Think of this, if the key is `0x112233`, the padded key is `0x11223300000000000000000000000000`. Then we find with the prefix `0x11223300`. This should find nothing. But if we don't have the `Where` part, it will match the padded key and return wrong results.